### PR TITLE
Add vhost-user-sound crate

### PR DESCRIPTION
**Pull request by epilys**

### CODE REVIEW

**Comment by stsquad** [comment]
Now this is re-based can we get this merged or at least reviewed?

**Comment by stefano-garzarella** [comment]
> Now this is re-based can we get this merged or at least reviewed?

Sorry, it has been very busy weeks, we will review by the end of this week.

File: crates/sound/src/device.rs [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+            let mut buffers = vec![];
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+            let message = Arc::new(IOMessage {
+                vring: vring.clone(),
+                status: VIRTIO_SND_S_OK.into(),
+                desc_chain: desc_chain.clone(),
+                descriptor: descriptors.last().cloned().unwrap(),
+            });
+            for (_, descriptor) in descriptors.iter().enumerate() {
+                match state {
```
Comment (MatiasVara) [inline crates/sound/src/device.rs]: Do we have the option of using a for-loop instead? I think it will be easier to read. We know that, in a request, the desc[0] is the header, desc[1..N-2] is the data, and desc[N-1] is the status, or I may be missing some corner-case.

File: crates/sound/src/device.rs [hunk 2]
```diff
@@ -0,0 +1,717 @@
...
+                        buffers.push(Buffer::new(buf, Arc::clone(&message)));
+                    }
+                }
+            }
+        }
+
+        if !stream_ids.is_empty() {
+            let b = audio_backend.write().unwrap();
+            for id in stream_ids {
+                b.write(id).unwrap();
```
Comment (MatiasVara) [inline crates/sound/src/device.rs]: If I understand correctly, in the tx queue, it collects all the data descriptors from a request. Then, It pushes for each descriptor that contains the data, the buffer and a reference to the message. When the status desc is identified, all the buffers are pushed back in the streams buffers queue. Then, for each `stream_id` that has been processed, it calls `write(stream_id)`. If you only have to notify guests after a `period_bytes` is consumed, would it not be better to only have a reference to the message for the set of buffers?

File: crates/sound/src/device.rs [hunk 3]
```diff
@@ -0,0 +1,717 @@
...
+        log::trace!("process_rx");
+        Ok(false)
+    }
+}
+
+pub struct VhostUserSoundBackend {
+    threads: Vec<RwLock<VhostUserSoundThread>>,
+    virtio_cfg: VirtioSoundConfig,
+    exit_event: EventFd,
+    audio_backend: RwLock<Box<dyn AudioBackend + Send + Sync>>,
```
Comment (MatiasVara) [inline crates/sound/src/device.rs]: Would it make sense to define `audio_backend` at `VhostUserSoundThread` instead? If that can't change over the life of the thread, it may be better to pass it on just once at creation. This would prevent changing the signature of `process_tx()` and `process_ctr()`. Also, shall we ensure internal mutability for the audio_backend? In that case, shall we re-define audio_backend as `Arc<Box<dyn AudioBackend + Send + Sync>>`?

File: crates/sound/src/stream.rs [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    Stop,
+}
+
+macro_rules! set_new_state {
+    ($new_state_fn:ident, $new_state:expr, $($valid_source_states:tt)*) => {
+        pub fn $new_state_fn(&mut self) -> Result<()> {
+            if !matches!(self, $($valid_source_states)*) {
+                return Err(Error::InvalidStateTransition(*self, $new_state));
+            }
+            *self = $new_state;
```
Comment (MatiasVara) [inline crates/sound/src/stream.rs]: If I understand correctly, here defines a macro to check if a cmd is allowed in a set of potential current states. For example, the `prepare` cmd is only allowed if the current state is `prepare`, `set_parameters` and `release`. If the cmd is executed in a different current state, the function fails. If the cmd is allowed based on the virtio spec, the state transits to the triggered command. This verification happens in the audio_backend itself (see https://github.com/epilys/vhost-device/blob/904391c35b83e885d6d36be48ccd9f72742785c9/crates/sound/src/audio_backends/alsa.rs#L432). In our case, we are doing it in the device before calling any `audio_backend` function. This is the purpose of the `get_stream_state()` method, i.e., to get the stream state and then verify that the cmd is allowed. AFAIU, the virtio spec does not tell what is the destination state during a transition but only the valid current state in which a cmd is allowed to be executed. The outgoing state may vary from backend to backend so the PCM state-machine for alsa may not be the same as the pipewire PCM-state machine although both conform to the virtio spec.

File: crates/sound/src/stream.rs [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+            Stop => {
+                write!(fmt, "VIRTIO_SND_R_PCM_STOP")
+            }
+        }
+    }
+}
+
+/// Internal state of a PCM stream of the VIRTIO Sound device.
+#[derive(Debug)]
+pub struct Stream {
```
Comment (MatiasVara) [inline crates/sound/src/stream.rs]: In our approach, we split the stream structure into information that is returned during GET_INFO and information that is used during SET_PARAM. For example, PCMParams is set during SET_INFO so we let the audio-backend handle that. We assume that the audio-backed will accept any configuration. Also, the state of a stream is stored in the audio-backend. We have extended the audio-backend interface to get the state of a stream by the `get_stream_state()` method. The device verifies that a command is executed in the right state. The audio backend decides the next state based on the spec. The device accesses the `stream.state` only for reading whereas the audio backend accesses it for writing and reading. This is because when a cmd is issued in a given state, different audio_backends may decide different transitions as far as they conform to the virtio spec.

**Comment by MatiasVara** [comment]
Hello @epilys and @stsquad , I just left some observations together with some questions to  help me better understand the code. Thanks.

File: crates/sound/src/device.rs (hunk 1 above)
Comment (epilys) [reply crates/sound/src/device.rs]: This is a for loop. Can you give me a written example of what you mean? In pseudocode etc

File: crates/sound/src/device.rs (hunk 2 above)
Comment (epilys) [reply crates/sound/src/device.rs]: > would it not be better to only have a reference to the message for the set of buffers?

The completion of a TX request can be considered an elapsed period notification but the reverse statement "tx requests must be completed (i.e. notify) after period_bytes are consumed" is not true.

In any case, the point of the references is reference count garbage collection; when all the buffers are dropped, the reference count of the IO message is zero so it is dropped as well, notifying the guest automatically.

File: crates/sound/src/stream.rs (hunk 4 above)
Comment (epilys) [reply crates/sound/src/stream.rs]: The state you change to is what "Possible valid transitions" in the spec means though, right?

File: crates/sound/Cargo.toml [hunk 6]
```diff
@@ -1,35 +1,37 @@
 [package]
 name = "vhost-user-sound"
 version = "0.1.0"
-authors = ["Stefano Garzarella <sgarzare@redhat.com>"]
+authors = ["Stefano Garzarella <sgarzare@redhat.com>", "Manos Pitsidianakis <manos.pitsidianakis@linaro.org>"]
 description = "A virtio-sound device using the vhost-user protocol."
 repository = "https://github.com/rust-vmm/vhost-device"
 readme = "README.md"
-keywords = ["vhost", "sound", "virtio-sound"]
+keywords = ["vhost", "sound", "virtio-sound", "virtio-snd", "vhost-user", "virtio"]
```
Comment (bilelmoussaoui) [inline crates/sound/Cargo.toml]: there is a limit of 5 keywords when publishing to crates.io iirc

File: crates/sound/rustfmt.toml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+edition = "2018"
```
Comment (bilelmoussaoui) [inline crates/sound/rustfmt.toml]:
```suggestion
edition = "2021"
```
maybe? but i believe this should be done in a separate PR so it can land already

File: crates/sound/src/audio_backends/null.rs [hunk 8]
```diff
@@ -1,30 +1,29 @@
...
 
-    fn read(&self, _stream_id: u32) -> Result<()> {
-        /*
-        let buf = req.data_slice().ok_or(Error::SoundReqMissingData)?;
-        let zero_mem = vec![0u8; buf.len()];
-
-        buf.copy_from(&zero_mem);
-
-        */
+    fn read(&self, _id: u32) -> Result<()> {
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/null.rs]:
```suggestion
    fn read(&self, id: u32) -> Result<()> {
```
as now this variable is "used"

File: crates/sound/src/lib.rs [hunk 9]
```diff
@@ -1,73 +1,196 @@
...
+use vhost_user_backend::{VhostUserDaemon, VringRwLock, VringT};
+use virtio_sound::*;
+use vm_memory::{
+    ByteValued, Bytes, GuestMemoryAtomic, GuestMemoryLoadGuard, GuestMemoryMmap, Le32,
+};
 
-use crate::vhu_sound::VhostUserSoundBackend;
+use crate::device::VhostUserSoundBackend;
 
+pub const SUPPORTED_FORMATS: u64 = 1 << VIRTIO_SND_PCM_FMT_U8
```
Comment (bilelmoussaoui) [inline crates/sound/src/lib.rs]: I wonder if those shouldn't be actually enums

File: crates/sound/src/lib.rs [hunk 10]
```diff
@@ -1,73 +1,196 @@
...
-    pub format: u8,
-    pub rate: u8,
+impl From<stream::Error> for Error {
+    fn from(val: stream::Error) -> Self {
+        Self::Stream(val)
+    }
+}
+
+#[derive(Debug)]
+pub struct InvalidControlMessage(u32);
```
Comment (bilelmoussaoui) [inline crates/sound/src/lib.rs]: needs a std::error::Error impl as well

File: crates/sound/src/lib.rs [hunk 11]
```diff
@@ -1,73 +1,196 @@
...
+    fn drop(&mut self) {
+        log::trace!(
+            "dropping ControlMessage {:?} reply = {}",
+            self.kind,
+            match self.code {
+                crate::virtio_sound::VIRTIO_SND_S_OK => "VIRTIO_SND_S_OK",
+                crate::virtio_sound::VIRTIO_SND_S_BAD_MSG => "VIRTIO_SND_S_BAD_MSG",
+                crate::virtio_sound::VIRTIO_SND_S_NOT_SUPP => "VIRTIO_SND_S_NOT_SUPP",
+                crate::virtio_sound::VIRTIO_SND_S_IO_ERR => "VIRTIO_SND_S_IO_ERR",
+                _ => "other",
```
Comment (bilelmoussaoui) [inline crates/sound/src/lib.rs]: if these were an enum, you would have an easier time here ;)

File: crates/sound/src/stream.rs [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+            *self = $new_state;
+            Ok(())
+        }
+    };
+}
+
+impl PCMState {
+    pub fn new() -> Self {
+        Self::SetParameters
+    }
```
Comment (bilelmoussaoui) [inline crates/sound/src/stream.rs]: Sounds like a Default impl ;)

**Review by bilelmoussaoui — COMMENTED** [review]
I am not familiar with virtio-sound stuff, but i left random comments on the rust part of it :)

File: crates/sound/src/audio_backends/null.rs (hunk 8 above)
Comment (epilys) [reply crates/sound/src/audio_backends/null.rs]: Used just for `log::trace!`, doesn't make any difference.

File: crates/sound/src/lib.rs (hunk 9 above)
Comment (epilys) [reply crates/sound/src/lib.rs]: They are actually not, they are `u8`.

File: crates/sound/src/lib.rs (hunk 10 above)
Comment (epilys) [reply crates/sound/src/lib.rs]: Anything that implements `std::fmt::{Debug,Display}` also implements `Error`. The `Error` trait has only provided methods, meaning you would only manually implement it to override the default methods.

File: crates/sound/src/lib.rs (hunk 11 above)
Comment (epilys) [reply crates/sound/src/lib.rs]: Doesn't seem difficult to me. But this trace print should be removed when upstreamed anyway.

File: crates/sound/src/lib.rs (hunk 10 above)
Comment (bilelmoussaoui) [reply crates/sound/src/lib.rs]: That is wrong.

Comment (bilelmoussaoui) [reply crates/sound/src/lib.rs]: The Error trait requires Debug + Display, so there can't be a blanket implementation for that.

Comment (epilys) [reply crates/sound/src/lib.rs]: Thanks, I misremembered. I will add an impl in a fixup commit

Comment (epilys) [reply crates/sound/src/lib.rs]: fixed

File: crates/sound/rustfmt.toml (hunk 7 above)
Comment (epilys) [reply crates/sound/rustfmt.toml]: crate (and others) is in 2018 edition, a follow up PR that updates them to 2021 makes sense

File: crates/sound/Cargo.toml (hunk 6 above)
Comment (epilys) [reply crates/sound/Cargo.toml]: fixed!

**Comment by epilys** [comment]
Force pushed with fixed suggestions by @bilelmoussaoui

File: crates/sound/src/device.rs (hunk 1 above)
Comment (MatiasVara) [reply crates/sound/src/device.rs]: Yes, it is.  I wonder if we can get rid of the "match state" statement like it is done at https://github.com/virtio-sound/vhost-device/blob/e500ee9576f616e00d38fefde779458b2df8a720/crates/sound/src/vhu_sound.rs#L506 There may have been some corner cases that I missed.

File: crates/sound/src/stream.rs (hunk 4 above)
Comment (MatiasVara) [reply crates/sound/src/stream.rs]: This is just to highlight that the next state could be decided for the audio backend and it could vary from audio backend to audio backend. I read the spec and crosvm code (and the code that I commented above) as "Possible valid transitions" are commands that can happen in a state. For example, 
```
PREPARE
The device prepares the stream (allocates resources, etc).
Possible valid transitions: set parameters, prepare, start, release.
```
This means that "set parameters", "prepare", "start" and "release" cmds are allowed in the PREPARE state. Conversely, the "prepare" cmd is only allowed in the SET PARAMETERS, PREPARE and RELEASE states. The spec leaves for the implementation to decide the next state. I think authors did it on purpose to let the audio backend to decide the transition. I submitted a change in the virtio spec regarding this to see if authors can clarify it (https://lists.oasis-open.org/archives/virtio-comment/202307/msg00158.html).

File: crates/sound/src/device.rs (hunk 1 above)
Comment (epilys) [reply crates/sound/src/device.rs]: Ah, I understand now. Personally the code you linked is really hard to read and understand. Whereas here the type and variant names are making the code more verbose without any performance penalty. Also, illegal states become unrepresentable.

File: crates/sound/src/stream.rs (hunk 4 above)
Comment (epilys) [reply crates/sound/src/stream.rs]: I don't understand the point you're trying to make though. The code here checks the valid transitions, and yes what the transition will be is the caller's decision.

Comment (stsquad) [reply crates/sound/src/stream.rs]: AIUI all this ensures is the backend doesn't attempt to transition to a state the spec doesn't allow.

Comment (epilys) [reply crates/sound/src/stream.rs]: @stsquad that is what the code this comment points to does, yes

Comment (MatiasVara) [reply crates/sound/src/stream.rs]: Thanks both for the clarification. This is a minor comment that we can discuss further at a later time. To illustrate my point, in the pipewire backend, we decided to stay in the `Self::Prepare` state if the `set_parameter` cmd is issued and the current state is `Self::Prepare`. It was simpler to do it like this. I think the PCM state machine shall be audio-backend-dependent. I may be wrong but I think the spec does not tell you what is the next state but only the cmds that are valid in a given current state. I will try to clarify this with the authors of the spec.

Comment (epilys) [reply crates/sound/src/stream.rs]: This code allows that behavior (staying from Prepare to Prepare), so **_I am still not understanding_** what your point is about. I think you have probably not understood yourself.

File: crates/sound/src/stream.rs (hunk 12 above)
Comment (epilys) [reply crates/sound/src/stream.rs]: forgot to comment, fixed in latest force push *

File: crates/sound/src/audio_backends/alsa.rs [hunk 13]
```diff
@@ -0,0 +1,555 @@
...
+                            msg.code = VIRTIO_SND_S_NOT_SUPP;
+                            continue;
+                        } else {
+                            st.params.features = request.features;
+                            st.params.buffer_bytes = request.buffer_bytes;
+                            st.params.period_bytes = request.period_bytes;
+                            st.params.channels = request.channels;
+                            st.params.format = request.format;
+                            st.params.rate = request.rate;
+                            st.params.period_bytes = request.period_bytes;
```
Comment (dorindabassey) [inline crates/sound/src/audio_backends/alsa.rs]: duplicated code?

Comment (epilys) [reply crates/sound/src/audio_backends/alsa.rs]: Indeed, thank you :) I pushed a fix.

**Comment by epilys** [comment]
Ping @stefano-garzarella what do you think about the state of the PR?

**Comment by stefano-garzarella** [comment]
> Ping @stefano-garzarella what do you think about the state of the PR?

Sorry, I was on PTO last week, and I'm catching up the last things.
I'll do a final look in the next days, but if @dorindabassey, @MatiasVara, and @bilelmoussaoui are happy, it is fine by me.

I just saw some questions marked resolved, but without any comments:
https://github.com/virtio-sound/vhost-device/pull/18#discussion_r1286933629
https://github.com/virtio-sound/vhost-device/pull/18#discussion_r1286935803

Maybe they are answered in some other comments that I'm missing.

File: crates/sound/src/device.rs (hunk 3 above)
Comment (epilys) [reply crates/sound/src/device.rs]: I don't see a reason to unless it's actually necessary by follow-up functionality, which would make it okay.

File: crates/sound/src/stream.rs (hunk 5 above)
Comment (epilys) [reply crates/sound/src/stream.rs]: If you look at alsa.rs, you will see that the audio backend is handling the state transitions here as well.

**Comment by epilys** [comment]
@stefano-garzarella oops thanks a lot! I thought I had answered it all. I left replies but everyone feel free to re-open anything unresolved or leave new comments.

File: crates/sound/src/device.rs [hunk 14]
```diff
@@ -0,0 +1,717 @@
...
+            let mut state = TxState::Ready;
+            let mut buffers = vec![];
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+            let message = Arc::new(IOMessage {
+                vring: vring.clone(),
+                status: VIRTIO_SND_S_OK.into(),
+                desc_chain: desc_chain.clone(),
+                descriptor: descriptors.last().cloned().unwrap(),
+            });
+            for (_, descriptor) in descriptors.iter().enumerate() {
```
Comment (bilelmoussaoui) [inline crates/sound/src/device.rs]: Is enumerate needed here if you don't need the index?

File: crates/sound/src/audio_backends/pipewire.rs [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+impl PwBackend {
+    pub fn new(streams: Arc<RwLock<Vec<Stream>>>) -> Self {
+        Self { streams }
+    }
+}
+
+impl AudioBackend for PwBackend {
+    fn write(&self, stream_id: u32) -> Result<()> {
+        log::trace!("PipewireBackend write stream_id {}", stream_id);
+        _ = std::mem::take(&mut self.streams.write().unwrap()[stream_id as usize].buffers);
```
Comment (dorindabassey) [inline crates/sound/src/audio_backends/pipewire.rs]: Hi Manos, I tried to do a rebase locally for compatibility purposes, and i noticed that when using `let st_buffer = std::mem::take(&mut self.streams.write().unwrap()[stream_id as usize].buffers);` it is constantly putting buffers in the variable `st_buffer` but when I use  `let st_buffer = &mut self.streams.write().unwrap()[stream_id as usize].buffers st_buffer` it is only populated with the first bytes of data at position 0, and the write fn seems to be called only once, I see that in the alsa backend you are using `&mut streams.write().unwrap()[stream_id]` could you explain to me the proper way to pass this buffer? in our case where I am rebasing [here](https://github.com/dorindabassey/vhost-device/commit/608538dd00805456ff6172ed976b306e196f3a34) , we  iterate over the Buffer elements to collect the bytes value into the data vector and then obtain a slice `& [u8]` from that data of `Vec<u8>` and cast to a ptr that can be used in the pipewire `spa_ringbuffer_write_data` callback. but the audio turn out to be choppy while using this `std::mem::take(&mut self.streams.write().unwrap()[stream_id as usize].buffers);` and i get no audio data for `let mut st_buffer = &mut self.streams.write().unwrap()[stream_id as usize].buffers` (probably because in this case the write fxn is called only once). in our write fxn [here](https://github.com/MatiasVara/vhost-device-for-virtio-sound/commit/bce65ee60662d3c6ea017ec78f9321689799c1ce) without the rebase, the audio that we play is not choppy, but they're other underlying issues of not being able to call speaker-test twice etc(unrelated). In summary, I want to understand how you pass the buffer around for alsa, and why the issues with handling this buffers in our backend. I would appreciate if you could take a look at the reference commits thank you :)

Comment (epilys) [reply crates/sound/src/audio_backends/pipewire.rs]: hello Dorinda :), the reason I'm doing `std::mem::take` here is because when the buffers are `Drop`ed the IO message reply is sent back  (from the `impl Drop` of the buffer).

Essentially do whatever you need to do with the buffers in the backend, *then* drop them.

 You can do a `&mut` borrow and write/read to the data buffer just fine, and when you want to mark the IO message as **finished** and **reply** to the guest, you make sure there's no living reference to any of the related buffers. Does that help?

File: crates/sound/src/device.rs (hunk 14 above)
Comment (epilys) [reply crates/sound/src/device.rs]: Doesn't seem like it, leftover from earlier versions or mistake. Will remove.

Comment (epilys) [reply crates/sound/src/device.rs]: Fixed, thanks!

**Comment by epilys** [comment]
Force pushed with fixes of style/code problems spotted by @bilelmoussaoui, + some Eq derives

File: crates/sound/src/audio_backends/alsa.rs [hunk 16]
```diff
@@ -0,0 +1,550 @@
...
+                virtio_sound::VIRTIO_SND_PCM_FMT_IEC958_SUBFRAME => Format::iec958_subframe(),
+                _ => Format::Unknown,
+            })?;
+
+            hwp.set_access(Access::RWInterleaved)?;
+
+            // > A period is the number of frames in between each hardware interrupt.
+            // - https://www.alsa-project.org/wiki/FramesPeriods
+
+            // FIXME
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/alsa.rs]: What are those FIXMEs are about? Should drop the parts that are no longer relevant or properly document what is needed to be fixed.

File: crates/sound/src/audio_backends/alsa.rs [hunk 17]
```diff
@@ -0,0 +1,550 @@
...
+                other => panic!("Invalid virtio-sound stream: {}", other),
+            },
+            false,
+        )?;
+
+        {
+            let hwp = HwParams::any(&pcm)?;
+            hwp.set_channels(s.params.channels.into())?;
+            hwp.set_rate(
+                match s.params.rate {
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/alsa.rs]: As this is something done also in the pipewire backend, i wonder if it really should be some enum or a wrapper around u8 with these constants and have a function that gives you the actual rate value.

File: crates/sound/src/audio_backends/alsa.rs [hunk 18]
```diff
@@ -0,0 +1,550 @@
...
+            stream.buffers.pop_front();
+        }
+    }
+    match mmap.status().state() {
+        State::Running => {
+            return Ok(false);
+        } // All fine
+        State::Prepared => {
+            //log::trace!("Starting audio output stream");
+            //p.start()?
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/alsa.rs]: Is it expected the Prepared/Suspended parts are commented?

File: crates/sound/src/audio_backends/alsa.rs [hunk 19]
```diff
@@ -0,0 +1,550 @@
...
+            //p.resume()?
+        }
+        n => panic!("Unexpected pcm state {:?}", n),
+    }
+    Ok(true) // Call us again, please, there might be more data to write
+}
+
+fn write_samples_io(
+    p: &alsa::PCM,
+
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/alsa.rs]:
```suggestion
```

File: crates/sound/src/audio_backends/alsa.rs [hunk 20]
```diff
@@ -0,0 +1,550 @@
...
+    stream: &mut Stream,
+    io: &mut alsa::pcm::IO<u8>,
+) -> AResult<bool> {
+    macro_rules! avail {
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/alsa.rs]: Why define a macro if it is only used once? :)

File: crates/sound/src/audio_backends/alsa.rs [hunk 21]
```diff
@@ -0,0 +1,550 @@
...
+        })?;
+        if written == 0 {
+            break;
+        };
+    }
+
+    match p.state() {
+        State::Suspended | State::Running => Ok(false), // All fine
+        State::Prepared => {
+            //p.start()?;
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/alsa.rs]: Commented code here as well

File: crates/sound/src/audio_backends/alsa.rs [hunk 22]
```diff
@@ -0,0 +1,550 @@
...
+            }
+        });
+
+        Self { sender }
+    }
+
+    fn run(
+        streams: Arc<RwLock<Vec<Stream>>>,
+        receiver: Receiver<AlsaAction>,
+    ) -> std::result::Result<(), Box<dyn std::error::Error>> {
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/alsa.rs]:
```suggestion
    ) -> Result<(), Box<dyn std::error::Error>> {
```
As you have a `AResult` / `CrateResult`  already.

Comment (epilys) [reply crates/sound/src/audio_backends/alsa.rs]: Fixed!

File: crates/sound/src/audio_backends/alsa.rs (hunk 21 above)
Comment (epilys) [reply crates/sound/src/audio_backends/alsa.rs]: Fixed/removed

File: crates/sound/src/audio_backends/alsa.rs (hunk 20 above)
Comment (epilys) [reply crates/sound/src/audio_backends/alsa.rs]: (It was used twice in a previous review iteration in the PR)

Fixed/removed

File: crates/sound/src/audio_backends/alsa.rs (hunk 19 above)
Comment (epilys) [reply crates/sound/src/audio_backends/alsa.rs]: Fixed/removed

File: crates/sound/src/audio_backends/alsa.rs (hunk 18 above)
Comment (epilys) [reply crates/sound/src/audio_backends/alsa.rs]: They were left over from the ALSA crate's example code, removed

File: crates/sound/src/audio_backends/alsa.rs (hunk 16 above)
Comment (epilys) [reply crates/sound/src/audio_backends/alsa.rs]: Wrote down what the FIXME was about and removed the commented code.

File: crates/sound/src/audio_backends/alsa.rs (hunk 17 above)
Comment (epilys) [reply crates/sound/src/audio_backends/alsa.rs]: Agreed. I think it's early/unnecessary to do it in this PR, but I might do it next time I force push

File: crates/sound/src/audio_backends/alsa.rs (hunk 16 above)
Comment (epilys) [reply crates/sound/src/audio_backends/alsa.rs]: (Feel free to resolve the conversation if the added comment is adequate)

**Comment by epilys** [comment]
Force pushed with fixes in alsa.rs.

File: crates/sound/src/audio_backends/alsa.rs (hunk 16 above)
Comment (bilelmoussaoui) [reply crates/sound/src/audio_backends/alsa.rs]: Apparently I can't resolve the conversation myself. But lgtm

**Review by bilelmoussaoui — APPROVED** [review]
lgtm from my side

File: crates/sound/src/audio_backends/pipewire.rs (hunk 15 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pipewire.rs]: sorry I don't understand. CMIIW, the reference st_buffer [here](https://github.com/dorindabassey/vhost-device/blob/608538dd00805456ff6172ed976b306e196f3a34/crates/sound/src/audio_backends/pipewire.rs#L219) goes out of scope at the end of each write function, and Rust automatically handles dropping the reference. what do you mean by to mark the IO message as finished and reply to the guest and how do i ensure that there is no living reference to
let mut st_buffer = &mut self.stream_params.write().unwrap()[stream_id as usize].buffers;

Comment (epilys) [reply crates/sound/src/audio_backends/pipewire.rs]: It's a reference counted resource. Each Buffer holds 1 reference. When a Buffer is dropped (at the end of a scope, or manually with `drop(_)`), it removes its reference from the count. When the count goes to 0, the IO message is dropped as well. 

> what do you mean by to mark the IO message as finished and reply to the guest

'Mark' was the wrong word to choose here, I meant when you are done with the IO message and want to complete it, all you have to do is drop the buffers.

**Review by stefano-garzarella — APPROVED** [review]

**Comment by stefano-garzarella** [comment]
@epilys I think we are ready to merge this PR ;-)
Can you do the last rebase?

About next steps, IIUC @MatiasVara and @dorindabassey are working on porting the pipewire backend from the old implementation to this new one.
@epilys do you plan to work on the `read` path?

**Review by dorindabassey — APPROVED** [review]

**Review by MatiasVara — APPROVED** [review]

**Comment by epilys** [comment]
@stefano-garzarella I wonder if we can force push this branch instead. My reasoning is that if the existing pipewire backend is ported to the PR, it'd be cleaner to apply it on top of it instead of introducing the extra steps of turning the existing pipewire code off to make compilation succeed and then replace it with new one. What do you think?

> @epilys do you plan to work on the read path?

Certainly, now that I did it in the qemu device too I have the knowledge to do a PR for it

**Comment by stefano-garzarella** [comment]
> @stefano-garzarella I wonder if we can force push this branch instead. My reasoning is that if the existing pipewire backend is ported to the PR, it'd be cleaner to apply it on top of it instead of introducing the extra steps of turning the existing pipewire code off to make compilation succeed and then replace it with new one. What do you think?

Since this is a dev repo, I think it should be fine.
I'll create a branch with the current main, in the case we will need it in the future, and force push this PR.

> 
> > @epilys do you plan to work on the read path?
> 
> Certainly, now that I did it in the qemu device too I have the knowledge to do a PR for it

Great ;-)

**Comment by stefano-garzarella** [comment]
Done, the previous `virtio-sound` branch is now [virtio-sound-v0.0](https://github.com/virtio-sound/vhost-device/tree/virtio-sound-v0.0) branch
